python/fusion_engine_client/parsers/mixed_log_reader.py

```diff
@@ -466,9 +488,27 @@ def filter_in_place(self, key, clear_existing: bool = False):
                 self.message_types = copy.deepcopy(self._original_message_types)
                 self.time_range = copy.deepcopy(self._original_time_range)
                 self.remove_invalid_p1_time = False
+                self.requested_source_ids = None
             else:
                 self.index = self._original_index
 
+        # Set requested source IDs.
+        if source_ids is not None:
+            source_ids = set(source_ids)
+            # Apply source IDs that are available.
+            if self.available_source_ids != source_ids:
+                unavailable_source_ids = list(source_ids.difference(self.available_source_ids))
+                if len(unavailable_source_ids) > 0:
+                    self.logger.warning('Not all source IDs requested are available. Cannot extract the following '
+                                        'source IDs: {}'.format(unavailable_source_ids))
+                source_ids = list(source_ids.intersection(self.available_source_ids))
+                if len(source_ids) == 0:
+                    raise ValueError("Requested source ID(s) unavailable. Exiting.")
```

Comment:
This message is a little confusing. First exceptions don't mean the application has to exit, that's not their decision. But also "unavailable" implies something within the class. Here this is simply an input error: when specifying a list of desired source IDs, it doesn't make sense for the list to be empty.

Something like "Requested source ID list cannot be empty." would be more clear.

Reply:
This is actually supposed to convey that all the source IDs requested by the user are unavailable!

Reply:
Ah I see, this is after the intersection. That's ok then, but we should probably just print a warning. Reading should work gracefully like normal, so it'll basically just output nothing. Same as if you gave it a time range that didn't cover the data. It shouldn't raise an exception.